### PR TITLE
Support for Flash Memory W25N01G of 128MB

### LINKED ISSUE #7885: No dataflash chip found on Diatone Mamba MK4 H743 FC
## Current Behavior
I have a new Diatone Mamba MK4 H743 FC that is supposed to have 128mb internal flash memory for blackbox logging.  I'm running INAV 4.1 (first time user of INAV), and it says in the top right corner that there is "No dataflash chip found".  Am I missing something as a new user?  Or is this a bug?

no data flash

- FC Board name and vendor: Diatone Mamba MK4 H743 FC
- INAV version string: 4.1

### LINKED ISSUE #8098: BlackBox: Internal flash memory W25N01G not supported
## Current Behavior

With amazement I noticed that on my ZEEZF7 V1.2 the 128 Mb internal flash for the BlackBox is not recognized.
With Betaflight, as you can see from the image below, no problem.

## Steps to Reproduce

1. Flash iNav V5.0 RC2 or 4.1, is indifferent
2. Connect to the FC, choose the default settings and go to the BlackBox section, you will see that it does not find the flash

## Suggested solution(s)

Is it possible that the target for the ZEEZF7 regarding the internal flash is wrong?
I checked it and the pins comparing with Betaflight seem to be declared correctly.

## Additional context

iNav: onboard flash not found.
immagine

Betaflight: 128 Mb onboard flash found.
immagine

---

INAV/ZEEZF7 5.0.0 Jun  3 2022 / 21:25:20 (46afd3a3)
GCC-10.2.1 20201103 (release)

### PR DESCRIPTION
**Pull request by JulioCesarMatias**
Along with @robustini and @hogthrob, I'm trying to bring this support to iNav, countless FC are currently using this flash memory. Closes https://github.com/iNavFlight/inav/issues/8098 and Closes https://github.com/iNavFlight/inav/issues/7885

- [x] Add the drivers
- [x] Make FC read multiple flash memories (added by @hogthrob).
- [x] Make W25N01G memory work regardless of whether M25P16 is set or not.
- [x] Fix FC crash if flash memory is not found (For some reason user removed from FC PCB, or it burned).
- [x] Make Reading and Writing work (fixed by @hogthrob).
- [x] Final Tests (Wainting...)
- [x] Working?

*Release Notes:*

Added support for 128MB W25N01G external flash memory, present in some new flight controllers. Discovered boards that use this chip: ZEEZF7 and MAMBAH743.

### CODE REVIEW

**Comment by robustini** [comment]
The dataflash is correctly detected, at the moment it is not possible to write or read.
I tested it with this [commit](https://github.com/iNavFlight/inav/commit/101e345b4f21afe2319c0235d0762082126a9d09) as well.

immagine

**Comment by DzikuVx** [comment]
That is awesome, thanks a lot for starting to work on it @JulioCesarMatias

**Comment by JulioCesarMatias** [comment]
@DzikuVx Can we add the `Release Notes` label to this PR?

**Comment by MrD-RC** [comment]
Can you add a section in the PR description with a release notes heading. Under that, write what should be included in the release notes. It just makes creating the releases easier.

**Comment by robustini** [comment]
@JulioCesarMatias thanks for the latest updates, i compiled with that PR but it still doesn't work.

https://user-images.githubusercontent.com/3864712/176109862-649e4334-b37a-4718-8226-78eb0b614ecf.mp4

**Comment by hogthrob** [comment]
@robustini : The code in my branch does erase successfully, so this is an indication that there is a problem in @JulioCesarMatias code. BTW, "erase" is not equal "writing", these are completely different operations. That one works doesn't imply the other does. 

However, in my branch (not @JulioCesarMatias PR which you tested), everything (read, write, erase) is working. 
Also blackbox recording (and display of amount of logged data aka free space) is working, although I "armed" the device for that.
Wasn't aware the logging also works unarmed.
 
Unfortunately, in my branch saving the blackbox log from flash is not working, though. I can see the log being written when I inspect the content using the low level `flash_read` cli command (#define USE_FLASH_TOOLS reguired in target.h) but the blackbox UI is not saving it for unknown reasons. 

@JulioCesarMatias code from this PR (not taking from the PR but his branch, though) does not produce working code for the MAMBAH743 target when I build this on my machine. That may not be the fault of the his changes but coming from the master branch but I didn't have the time to check this.

**Comment by robustini** [comment]
> @robustini : The code in my branch does erase successfully, so this is an indication that there is a problem in @JulioCesarMatias code. BTW, "erase" is not equal "writing", these are completely different operations. That one works doesn't imply the other does.
> 
> However, in my branch (not @JulioCesarMatias PR which you tested), everything (read, write, erase) is working. Also blackbox recording (and display of amount of logged data aka free space) is working, although I "armed" the device for that. Wasn't aware the logging also works unarmed.
> 
> Unfortunately, in my branch saving the blackbox log from flash is not working, though. I can see the log being written when I inspect the content using the low level `flash_read` cli command (#define USE_FLASH_TOOLS reguired in target.h) but the blackbox UI is not saving it for unknown reasons.
> 
> @JulioCesarMatias code from this PR (not taking from the PR but his branch, though) does not produce working code for the MAMBAH743 target when I build this on my machine. That may not be the fault of the his changes but coming from the master branch but I didn't have the time to check this.

yes, I get confused with Betaflight where BlackBox is configurable in ALWAYS and logs even when unarmed.

**Comment by robustini** [comment]
Tested now with the latest fix and flashgo, it hangs on reading them:

immagine

**Comment by DzikuVx** [comment]
@JulioCesarMatias when ready to merge, please squassh to a single commit if possible. I will want to cherry-pick it to 5.1 branch

**Comment by hogthrob** [comment]
I pushed my working code to my repository https://github.com/hogthrob/inav/tree/flash_w25n01gSupport , see discussion in  #8098 (work is based on this branch). No need to create an additional PR, let's wait for the report of @robustini .

BTW, still fighting to find the correct formatting setting for VS Code to match iNav formatting exactly style.

**Comment by JulioCesarMatias** [comment]
> I pushed my working code to my repository https://github.com/hogthrob/inav/tree/flash_w25n01gSupport , see discussion in #8098 (work is based on this branch). No need to create an additional PR, let's wait for the report of @robustini .
> 
> BTW, still fighting to find the correct formatting setting for VS Code to match iNav formatting exactly style.

Very nice @hogthrob, this PR owes all the credits to you! I passed the changes to my branch, could you test it to make sure I didn't miss anything? >> https://drive.google.com/file/d/16iB-P9WEMapCN6pCEl2jhhyyRF_O3oPi/view?usp=sharing

**Comment by JulioCesarMatias** [comment]
@hogthrob About the "dummy" flash you added, I preferred not to keep it, it is interesting yes, but visually it pollutes the algorithm... I think just checking that there is no flash memory on the FC PCB (even defined in target.h) , just indicating to the algorithm that it was not found is enough.

**Comment by hogthrob** [comment]
There is a couple of places outside flash.c  calling various functions from flash.c are without checking if the flash was detected. Having such a dummy is the easiest way to get around this until this problem has been mitigated.

The other solution is to guard  every relevant call with along the lines `if (flash_not_found) { return default_answer; } else { flash->,,, }` . One could also call this pollution. 

But honestly, I don't care as long as the resulting code works, It is "your" project and your PR.

**Comment by JulioCesarMatias** [comment]
@hogthrob Did the algorithm compiled through my branch work?

**Comment by robustini** [comment]
> I pushed my working code to my repository https://github.com/hogthrob/inav/tree/flash_w25n01gSupport , see discussion in #8098 (work is based on this branch). No need to create an additional PR, let's wait for the report of @robustini .
> 
> BTW, still fighting to find the correct formatting setting for VS Code to match iNav formatting exactly style.

Doesn't see the dataflash here.
I attach the modified target.c for my board, it should be correct.

[target.zip](https://github.com/iNavFlight/inav/files/9032565/target.zip)

**Comment by robustini** [comment]
> I pushed my working code to my repository https://github.com/hogthrob/inav/tree/flash_w25n01gSupport , see discussion in #8098 (work is based on this branch). No need to create an additional PR, let's wait for the report of @robustini .
> 
> BTW, still fighting to find the correct formatting setting for VS Code to match iNav formatting exactly style.

Doesn't see the dataflash here.
Hope my target.h is correct.

[target.zip](https://github.com/iNavFlight/inav/files/9032572/target.zip)

**Comment by robustini** [comment]
> I pushed my working code to my repository https://github.com/hogthrob/inav/tree/flash_w25n01gSupport , see discussion in #8098 (work is based on this branch). No need to create an additional PR, let's wait for the report of @robustini .
> 
> BTW, still fighting to find the correct formatting setting for VS Code to match iNav formatting exactly style.

I compile in and I report the only anomaly.
Excuse the perhaps trivial question but in iNav there is the possibility to write on the blackbox unarmed as with Betaflight?
Here at home at the moment I don't have tx and battery and I can't test the writes in the dataflash if it doesn't do it unarmed.

immagine

**Comment by robustini** [comment]
@hogthrob no, there is still something wrong, it hangs here.

immagine

**Comment by stronnag** [comment]
Couple of observations, apologies if they are trivial / stuff you already know / have tried.

* flashgo  has a hard coded read size of 4KB; would it help if this were configurable ?
* If you put blackbox on a switch, does that get around the arming requirement ?
* Can you see / read the flash in `MSC` mode ?
* FYI, mwptools includes a [CLI based flash downloader]( https://github.com/stronnag/mwptools/blob/master/src/samples/flash_dump.rb), if that's any automation help

**Comment by robustini** [comment]
> Couple of observations, apologies if they are trivial / stuff you already know / have tried.
> 
>     * flashgo  has a hard coded read size of 4KB; would it help if this were configurable ?

It is probable

> 
>     * If you put blackbox on a switch, does that get around the arming requirement ?

No

> 
>     * Can you see / read the flash in `MSC` mode ?

No, it recognizes a device without a driver

> 
>     * FYI, mwptools includes a [CLI based flash downloader](https://github.com/stronnag/mwptools/blob/master/src/samples/flash_dump.rb), if that's any automation help

**Comment by stronnag** [comment]
FWIW, now you can specify the read size, don't think it will help you:

```
$ flashgo -test
Using /dev/ttyACM0
Read buffer size 4096 bytes
Firmware: INAV
Version: 5.0.0
Entering test mode for 2097152b
Data flash 2097152 / 2097152 (100%)
Downloading to bbl_2022-07-02_093606.TXT
[▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
2097152 bytes in 38.5s, 54521.0 bytes/s
$ flashgo -test -readsize 8K
Using /dev/ttyACM0
Read buffer size 8192 bytes
Firmware: INAV
Version: 5.0.0
Entering test mode for 2097152b
Data flash 2097152 / 2097152 (100%)
Downloading to bbl_2022-07-02_093705.TXT
[▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
2097152 bytes in 39.1s, 53651.3 bytes/s
$ flashgo -test -readsize 1K
Using /dev/ttyACM0
Read buffer size 1024 bytes
Firmware: INAV
Version: 5.0.0
Entering test mode for 2097152b
Data flash 2097152 / 2097152 (100%)
Downloading to bbl_2022-07-02_093758.TXT
[▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
2097152 bytes in 49.3s, 42499.4 bytes/s
```

**Comment by robustini** [comment]
> FWIW, now you can specify the read size, don't think it will help you:
> 
> ```
> $ flashgo -test
> Using /dev/ttyACM0
> Read buffer size 4096 bytes
> Firmware: INAV
> Version: 5.0.0
> Entering test mode for 2097152b
> Data flash 2097152 / 2097152 (100%)
> Downloading to bbl_2022-07-02_093606.TXT
> [▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
> 2097152 bytes in 38.5s, 54521.0 bytes/s
> $ flashgo -test -readsize 8K
> Using /dev/ttyACM0
> Read buffer size 8192 bytes
> Firmware: INAV
> Version: 5.0.0
> Entering test mode for 2097152b
> Data flash 2097152 / 2097152 (100%)
> Downloading to bbl_2022-07-02_093705.TXT
> [▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
> 2097152 bytes in 39.1s, 53651.3 bytes/s
> $ flashgo -test -readsize 1K
> Using /dev/ttyACM0
> Read buffer size 1024 bytes
> Firmware: INAV
> Version: 5.0.0
> Entering test mode for 2097152b
> Data flash 2097152 / 2097152 (100%)
> Downloading to bbl_2022-07-02_093758.TXT
> [▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇] 2.0MB/2.0MB 100% 0s
> 2097152 bytes in 49.3s, 42499.4 bytes/s
> ```

I tried but the problem remains, it doesn't depend on flashgo.

**Comment by stronnag** [comment]
> I tried but the problem remains, it doesn't depend on flashgo.

I know, just though the option to at least easily try different read sizes might be useful.

**Comment by hogthrob** [comment]
@robustini: target.h looks ok on the first glance. In the iNAV Configurator 'cli' what is the output of `flash_info` ?

**Comment by robustini** [comment]
> @robustini: target.h looks ok on the first glance. In the iNAV Configurator 'cli' what is the output of `flash_info` ?

Flash sectors=1024, sectorSize=131072, pagesPerSector=64, pageSize=2048, totalSize=134217728
Paritions:
  0: BBMGMT    1003 1023
  1: FLASHFS   0 1002
FlashFS size=131465216, usedSize=0

**Comment by hogthrob** [comment]
@robustini : That looks good. Flash is being detected properly. Now next thing to test is `flash_write 0 IwasHere`. Followed by `flash_write 2048 X`. Now `flash_read 0 16` should show you `IwasHere` plus 8 time a repeating character. 
With that correct operation of flash is confirmed. The rest is "just" setup. BTW, I haven't tried `flashgo` so maybe it is not working. But I am just now looking at my blackbox logs from some testflights I did earlier today.

**Comment by robustini** [comment]
> flash_read 0 16


Entering CLI Mode, type 'exit' to return, or 'help'

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 2048 X
Wrote 2 bytes at 2048.

# flash_read 0 16
Reading 16 bytes at 0:
 IwasHereÿÿÿÿÿÿÿ

**Comment by robustini** [comment]
> @robustini : That looks good. Flash is being detected properly. Now next thing to test is `flash_write 0 IwasHere`. Followed by `flash_write 2048 X`. Now `flash_read 0 16` should show you `IwasHere` plus 8 time a repeating character. With that correct operation of flash is confirmed. The rest is "just" setup. BTW, I haven't tried `flashgo` so maybe it is not working.
> 
> But I am just now looking at my blackbox logs from some testflights I did earlier today, works for me.

`
Entering CLI Mode, type 'exit' to return, or 'help'

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 0 IwasHere
Wrote 9 bytes at 0.

# flash_write 2048 X
Wrote 2 bytes at 2048.

# flash_read 0 16
Reading 16 bytes at 0:
 IwasHereÿÿÿÿÿÿÿ
`

**Comment by hogthrob** [comment]
@robustini :That confirms flash is working as it should. If you now reboot the flightcontroller, and run `flash_info` again, you should see the useSize increased to 2048.  And `flash_read 0 16` should still give you the same output.

All is left to go out and fly.

**Comment by robustini** [comment]
@hogthrob very good, great job, thanks!
I repeat that however it would be useful to be able to log in even unarmed, I hope @DzikuVx can consider it as a feature.

**Comment by JulioCesarMatias** [comment]
@robustini Can you test the Hex I compiled? >> https://drive.google.com/file/d/15-2fg1rZk6kw74PPS6gs8UjfcI5o0Scn/view?usp=sharing

**Comment by robustini** [comment]
> @robustini Can you test the Hex I compiled? >> https://drive.google.com/file/d/15-2fg1rZk6kw74PPS6gs8UjfcI5o0Scn/view?usp=sharing

Good job mate! With your hex the MSC command work fine!
I can't test from CLI because you haven't put define for flash tools.

Flash sectors=1024, sectorSize=131072, pagesPerSector=64, pageSize=2048, totalSize=134217728
Paritions:
  0: FLASHFS   0 1023
FlashFS size=134217728, usedSize=0


immagine

File: src/main/drivers/flash.c [hunk 1]
```diff
@@ -30,92 +30,118 @@
...
 }
 
 bool flashIsReady(void)
 {
-#ifdef USE_FLASH_M25P16
-    return m25p16_isReady();
-#endif
-    return false;
+    // prevent the machine cycle from crashing if there is no external flash memory
+    if (!flash->detected) {
```
Comment (hogthrob) [inline src/main/drivers/flash.c]: This is not a perfectly safe approach to test for absence of flash, as this requires flashDeviceInit to have initialized the `flash` structure pointer before this check can be used. 
It also fails if flashfs (USE_FLASHFS) and none of the flash drivers are configured.
Easiest solution to this is to remove the `detected` field from the flash structure and put this as a static variable into the module flash.c and initialized with `false`. 
Added benefit of removing `detected` is that the flash structs array for the drivers could go to read-only memory (not RAM) if const is added to array definition as the function pointers do not change during run-time.

**Review by hogthrob — CHANGES_REQUESTED** [review]

**Comment by JulioCesarMatias** [comment]
@hogthrob what were the results after your commit? >> https://github.com/hogthrob/inav/commit/1bfada9c03c77ab064549b6f00b493152c9655a3

File: src/main/drivers/flash.c (hunk 1 above)
Comment (hogthrob) [reply src/main/drivers/flash.c]: An alternative and probably the easiest solution is to use as test `if (flash == NULL)` as the flash variable only gets assigned if a driver was successfully initialized.

**Comment by JulioCesarMatias** [comment]
@DzikuVx I don't know how to squash through Git, I made some attempts, but without success, and ended up compiling some things (such as doubling the number of commits of this PR)... But you can squash through the PR merge button

github-close-pull-request-with-squash

**Comment by JulioCesarMatias** [comment]
@robustini I had some problems with the brach when trying to do a squash, and I had to edit some files manually again... so to make sure this is working could you test it again? >> https://drive.google.com/file/d/1KzXKr64IKQ42YMsrYYa9UenF6rkIWQQI/view?usp=sharing

**Comment by robustini** [comment]
> @robustini I had some problems with the brach when trying to do a squash, and I had to edit some files manually again... so to make sure this is working could you test it again? >> https://drive.google.com/file/d/1KzXKr64IKQ42YMsrYYa9UenF6rkIWQQI/view?usp=sharing

It seems ok but I repeat it, at the moment I have not tested the writing and reading of a log file not being able to do it unarmed.

**Comment by robustini** [comment]
I am now detecting a strange thing, the serialpassthrough command in CLI does not exist.
Having compiled from the master branch by merging this PR by chance in 6.0.0 this command from the CLI was removed?

**Comment by robustini** [comment]
I'm trying to understand why my rx with ExpressLRS (CRFS protocol) with this firmware no longer communicates with the FC.
I flashed the official 5.0.0 it works correctly.
I had no problems when compiling the firmware.
I used the master branch plus this PR, nothing else.

**Comment by robustini** [comment]
Tests completed, this is the feedback.
The CRSF problem with my board does not depend on this PR but present on the current master (6.0.0), I don't know if something has changed on the management of the rx, or something is broken.
However it is not a problem related to this PR.

**Comment by robustini** [comment]
The compilation conflict is related to the CMakeLists.txt, "project (INAV VERSION 6.0.0)" in this PR while the branch version is 5.1.0.

**Comment by robustini** [comment]
CRFS problem fixed with #8191.

**Comment by JulioCesarMatias** [comment]
Okay, thanks to everyone involved. This work is ready to be merged!

**Pull request merged**